searx/network/context.py

```diff
@@ -0,0 +1,404 @@
+        timeout: Optional[float],
+    ):
+        self._retries: int = retries
+        # wrap http_client_factory here, so we can forget about this wrapping
+        self._http_client_factory = _TimeHTTPClientWrapper.wrap_factory(http_client_factory, self)
+        self._http_client: Optional[ABCHTTPClient] = None
+        self._start_time: float = start_time or default_timer()
+        self._http_time: float = 0.0
+        self._timeout: Optional[float] = timeout
+
+    @abstractmethod
+    def call(self, func: Callable[P, R], *args: P.args, **kwargs: P.kwargs) -> R:
+        """Call func within the network context.
+
+        The retry policy might call func multiple times.
+
+        Within the function self.get_http_client() returns an HTTP client to use.
+
+        The retry policy might send multiple times the same HTTP request
+        until it works or the retry count falls to zero.
+        """
+
+    @final
+    def request(self, *args, **kwargs):
+        """Convenient method to wrap a call to request inside the call method.
+
+        Use a new HTTP client to wrap a call to the request method using self.call
+        """
+        self._set_http_client()
+        return self.call(self._get_http_client().request, *args, **kwargs)
+
+    @final
+    def stream(self, *args, **kwargs):
+        """Convenient method to wrap a call to stream inside the call method.
+
+        Use a new HTTP client to wrap a call to the stream method using self.call
+        """
+        self._set_http_client()
+        return self.call(self._get_http_client().stream, *args, **kwargs)
+
+    @final
+    def get_http_runtime(self) -> Optional[float]:
+        """Return the amount of time spent on HTTP requests"""
+        return self._http_time
+
+    @final
+    def get_remaining_time(self, _override_timeout: Optional[float] = None) -> float:
+        """Return the remaining time for the context.
+
+        _override_timeout is not intended to be used outside this module.
+        """
+        timeout = _override_timeout or self._timeout or DEFAULT_TIMEOUT
+        timeout += 0.2  # overhead
+        timeout -= default_timer() - self._start_time
+        return timeout
+
+    @final
+    def _get_http_client(self) -> ABCHTTPClient:
+        """Return the HTTP client to use for this context."""
+        if self._http_client is None:
+            raise ValueError("HTTP client has not been set")
+        return self._http_client
+
+    @final
+    def _set_http_client(self):
+        """Ask the NetworkContext to use another HTTP client using the factory.
+
+        Use the method _get_new_client_from_factory() to call the factory,
+        so the NetworkContext implementations can wrap the ABCHTTPClient.
+        """
+        self._http_client = self._get_new_client_from_factory()
+
+    @final
+    def _reset_http_client(self):
+        self._http_client = None
+
+    def _get_new_client_from_factory(self):
+        return self._http_client_factory()
```

Comment:
> do we really need such a lot of indirections .. where private methods get/set private members?

I don't remember exactly the last audio call, may be I said it was private methods ?
I should have said protected methods rather than private methods.

> the reader needs to know that a private method / a method with an underline prefix 

it seems to be a common practice:
* a single `_` as prefix for protected methods
* a double `__` as prefix for private methods

See:
* https://softwareengineering.stackexchange.com/questions/228216/what-is-the-historical-reason-why-python-uses-the-double-underscore-for-class-pr
* https://discuss.python.org/t/private-protected-modifier-and-notation/22356
* https://stackoverflow.com/questions/6930144/underscore-vs-double-underscore-with-variables-and-methods

About indirection: I can't say I'm happy with them.
`_get_new_client_from_factory` is override by the different implementations:
https://github.com/searxng/searxng/blob/ffcff586939a548c6cd9b279e2d0badea55a6656/searx/network/context.py#L267-L268

https://github.com/searxng/searxng/blob/ffcff586939a548c6cd9b279e2d0badea55a6656/searx/network/context.py#L315-L316

https://github.com/searxng/searxng/blob/ffcff586939a548c6cd9b279e2d0badea55a6656/searx/network/context.py#L369-L370

Reading the code now, I think:
* `_set_http_client` can be not final, and just set `self._http_client`
* `_get_new_client_from_factory` can be removed.

So for example:

```python
class NetworkContextRetryFunction(NetworkContext):
    def call(self, func: Callable[P, R], *args: P.args, **kwargs: P.kwargs) -> R:
        ...

    def _set_http_clientself):
        self._http_client = _RetryFunctionHTTPClient(super()._get_new_client_from_factory(), self)
```